roms_tools/plot.py

```diff
@@ -597,13 +631,464 @@ def _add_field_to_ax(
+        title = field.long_name
+    else:
+        title = ""
+
+    # Slice the field horizontally as desired
+    def _slice_along_dimension(field, title, dim_name, idx):
+        field = field.sel(**{dim_name: idx})
+        title = title + f", {dim_name} = {idx}"
+        return field, title
+
+    if eta is not None:
+        field, title = _slice_along_dimension(field, title, horizontal_dims["eta"], eta)
+    if xi is not None:
+        field, title = _slice_along_dimension(field, title, horizontal_dims["xi"], xi)
+    if s is not None:
+        field, title = _slice_along_dimension(field, title, "s_rho", s)
+
+    # Compute layer and interface depth if necessary
+    compute_layer_depth = ("s_rho" in field.dims) or (
+        depth_contours and s is not None and lat is None and lon is None
+    )
+    compute_interface_depth = (
+        len(field.dims) == 2
+        and "s_rho" in field.dims
+        and layer_contours
+        and (lat is None)
+        and (lon is None)
+        and (depth is None)
+    )
+
+    def _compute_and_slice_depth(depth_type):
+        depth = compute_depth_coordinates(
+            grid_ds, zeta, depth_type=depth_type, location=loc, eta=eta, xi=xi
+        )
+        if not include_boundary:
+            depth = depth.isel(
+                **{
+                    dim: s
+                    for dim, s in slice_dict.get(loc, {}).items()
+                    if dim in depth.dims
+                }
+            )
+        if s is not None:
+            depth = depth.isel(s_rho=s)
+        return depth.load()
+
+    if compute_layer_depth:
+        layer_depth = _compute_and_slice_depth("layer")
+
+    if compute_interface_depth:
+        interface_depth = _compute_and_slice_depth("interface")
+
+    # Regrid laterally
+    if lat is not None or lon is not None:
+        if lat is not None:
+            lats = [lat]
+            title = title + f", lat = {lat}°N"
+        else:
+            resolution = infer_nominal_horizontal_resolution(grid_ds)
+            lats = _generate_coordinate_range(
+                field.lat.min().values, field.lat.max().values, resolution
+            )
+        lats = xr.DataArray(lats, dims=["lat"], attrs={"units": "°N"})
+
+        if lon is not None:
+            lons = [lon]
+            title = title + f", lon = {lon}°E"
+        else:
+            resolution = infer_nominal_horizontal_resolution(grid_ds, lat)
+            lons = _generate_coordinate_range(
+                field.lon.min().values, field.lon.max().values, resolution
+            )
+        lons = xr.DataArray(lons, dims=["lon"], attrs={"units": "°E"})
+
+        target_coords = {"lat": lats, "lon": lons}
+        lateral_regrid = LateralRegridFromROMS(field, target_coords)
+        field = lateral_regrid.apply(field).squeeze()
+
+        if compute_layer_depth:
+            layer_depth = lateral_regrid.apply(layer_depth).squeeze()
+
+    # Assign depth as coordinate
+    if compute_layer_depth:
```

Comment:
Consider pulling the layer depth computations out into their own method to reduce the overall size of this one (~300 LoC) and enable testing.

Reply:
Agreed. Let's leave it for a future PR?